ext/app/electron/WindowManager.ts

```diff
@@ -0,0 +1,95 @@
+import * as electron from "electron";
+import * as path from "path";
+import { GristLoadConfig, decodeUrl } from "app/common/gristUrls";
+import { BrowserWindow } from "electron";
+import { ElectronLoginSystem } from "./LoginSystem";
+
+export class WindowManager {
+
+  // null represents "not associated with a document".
+  private docToWindowMap: Map<string, BrowserWindow> = new Map();
+  private windowToDocMap: Map<BrowserWindow, string | null> = new Map();
+
+  constructor(
+    private gristConfig: GristLoadConfig,
+    private resolveDocId: (urlIdOrDocId: string) => Promise<string>
+  ) {}
+
+  public get(docId: string): BrowserWindow | null {
+    return this.docToWindowMap.get(docId) ?? null;
+  }
+
+  public add(docId: string | null): BrowserWindow {
+
+    if (docId) {
+      const win = this.docToWindowMap.get(docId);
+      if (win) {
+        return win;
+      }
+    }
+
+    const win = new BrowserWindow({
+      width: 1024,
+      height: 768,
+      webPreferences: {
+        nodeIntegration: false,
+        preload: path.join(__dirname, "preload.js"),
+        webviewTag: true
+      },
+      backgroundColor: "#42494B",
+      autoHideMenuBar: false,
+    });
+
+    if (docId) {
+      this.docToWindowMap.set(docId, win);
+    }
+    this.windowToDocMap.set(win, docId);
+
+    win.webContents.on("did-navigate", async (_, url) => {
+      const gristUrl = decodeUrl(this.gristConfig, new URL(url));
+      const oldDocId = this.windowToDocMap.get(win);
+      // Most of the time this would be a doc ID, but it could also be an URL ID.
+      const newDocIdOrUrlId = gristUrl.doc ?? null;
+      // oldDocId must exist because windowToDocMap knows about all existing windows.
+      this.docToWindowMap.delete(oldDocId!);
+      if (newDocIdOrUrlId) {
```

Comment:
This could potentially be null, looking at lines 44 + 46. Specifically if we have a window that wasn't previously on a doc.

In practice, I don't think this will break anything, but I don't think it's a correct type-system override either, so we should fix it.

Reply:
I must have missed this somehow. Thanks for pointing it out!